### PR TITLE
Move debug.console to a dedicated plugin

### PR DESCRIPTION
**Pull request by le-jeu**
Current `window.debug` breaks some plugins on mobile app with force desktop enable.
This is due to the override of `window.console` by an incomplete console object implementing log/warn/error only while plugin may use any Console api (debug/assert among others).

This PR removes this API from the core of IITC since it duplicates a working feature on IITC-Mobile.
It's replaced by a plugin version of the same feature, fixing the incomplete console API, so it could be use with devices that do not provide a debug console (say, IITC-Mobile for iOS)

### CODE REVIEW

File: core/code/chat.js [hunk 1]
```diff
@@ -811,23 +811,13 @@ window.chat.postMsg = function() {
   if(c == 'alerts')
     return alert("Jarvis: A strange game. The only winning move is not to play. How about a nice game of chess?\n(You can't chat to the 'alerts' channel!)");
 
+  // unknown tab, ignore
+  if (c != 'all' && c != 'faction')
```
Comment (johnd0e) [inline core/code/chat.js]: !==

File: core/code/debugging.js [hunk 2]
```diff
@@ -1,106 +1,21 @@
...
 
-  function overwrite(which) {
-    window.console[which] = function() {
-      nativeConsole[which].apply(nativeConsole, arguments);
-      window.debug.console[which].apply(window.debug.console, arguments);
-    }
-  }
+var noop = function() {};
+window.debug.renderDetails = noop;
+window.debug.printStackTrace = noop;
```
Comment (johnd0e) [inline core/code/debugging.js]: Where they used?

File: core/code/debugging.js [hunk 3]
```diff
@@ -1,106 +1,21 @@
...
 
-  overwrite("log");
-  overwrite("warn");
-  overwrite("error");
-}
+window.debug.console = function() {}
```
Comment (johnd0e) [inline core/code/debugging.js]: Is it for legacy plugins?

Comment (le-jeu) [reply core/code/debugging.js]: yes, all declarations are for legacy plugins

File: core/code/debugging.js (hunk 2 above)
Comment (le-jeu) [reply core/code/debugging.js]: no idea, I suppose they don't, but I was told not to break the API ^^

File: core/code/debugging.js (hunk 3 above)
Comment (johnd0e) [reply core/code/debugging.js]: Then I have objections - any user of legacy plugin can enable `debug-console` plugin by himself.

File: core/code/chat.js [hunk 4]
```diff
@@ -811,23 +811,13 @@ window.chat.postMsg = function() {
   if(c == 'alerts')
     return alert("Jarvis: A strange game. The only winning move is not to play. How about a nice game of chess?\n(You can't chat to the 'alerts' channel!)");
 
+  // unknown tab, ignore
+  if (c !== 'all' && c !== 'faction')
+    return;
```
Comment (johnd0e) [inline core/code/chat.js]: 'one true brace-style'

**Review by johnd0e — CHANGES_REQUESTED** [review]

File: core/code/chat.js [hunk 5]
```diff
@@ -590,7 +590,9 @@ window.chat.request = function() {
 // loads more if not.
 window.chat.needMoreMessages = function() {
   var activeTab = chat.getActive();
-  if(activeTab === 'debug') return;
+  if(activeTab !== 'all' && activeTab !== 'faction' && activeTab !== 'alerts') {
```
Comment (johnd0e) [inline core/code/chat.js]:
```suggestion
  if (activeTab !== 'all' && activeTab !== 'faction' && activeTab !== 'alerts') {
```

Comment (johnd0e) [reply core/code/chat.js]: Or `['all', 'faction', 'alerts'].indexOf(activeTab)`

File: plugins/debug-console.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    window.console[which] = function() {
+      if (nativeConsole) nativeConsole[which].apply(nativeConsole, arguments);
+      debugTab.console[which].apply(debugTab.console, arguments);
+    }
+  }
+
+  overwrite("log");
+  overwrite("warn");
+  overwrite("error");
+  overwrite("debug");
```
Comment (johnd0e) [inline plugins/debug-console.js]: single braces
Or `['log', 'warn', 'error', 'debug'].forEach(overwrite);`

File: plugins/debug-console.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  debugTab.console.renderLine('debug', arguments);
+}
+
+function overwriteNative() {
+  var nativeConsole = window.console;
+  window.console = L.extend({}, window.console);
+
+  function overwrite(which) {
+    window.console[which] = function() {
+      if (nativeConsole) nativeConsole[which].apply(nativeConsole, arguments);
```
Comment (johnd0e) [inline plugins/debug-console.js]: I'd prefer 1tbs, as line is long.

File: plugins/debug-console.js [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+  overwrite("log");
+  overwrite("warn");
+  overwrite("error");
+  overwrite("debug");
+}
+
+function setupPosting() {
+  if (!window.isSmartphone()) {
+    $('#chatinput input').keydown(function(event) {
+      var kc = (event.keyCode ? event.keyCode : event.which);
```
Comment (johnd0e) [inline plugins/debug-console.js]:
```suggestion
      var kc = event.keyCode ? event.keyCode : event.which;
```

File: plugins/debug-console.js [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+      if(kc === 13) { // enter
+        if ($('#chatcontrols .active').text() == 'debug') {
```
Comment (johnd0e) [inline plugins/debug-console.js]: eqeqeq

File: plugins/debug-console.js [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  $('#chatinput').submit(function(event) {
+    event.preventDefault();
+    userInput();
+  });
+}
+
+function userInput() {
+  if ($('#chatcontrols .active').text() != 'debug')
```
Comment (johnd0e) [inline plugins/debug-console.js]: eqeqeq

File: plugins/debug-console.js [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    return;
+
+  var msg = $.trim($('#chatinput input').val());
+  if(!msg || msg === '') return;
```
Comment (johnd0e) [inline plugins/debug-console.js]:
```suggestion
  if (!msg) { return; }
```

File: plugins/debug-console.js [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+  $('#chatcontrols').append('<a>debug</a>');
+  $('#chatcontrols a:last').click(debugTab.console.show);
+  $('#chat').append('<div style="display: none" id="debugconsole"><table></table></div>');
+
+  setupPosting();
+
+  if (window.useAndroidPanes()) {
+    android.addPane('debug', 'Debug', 'ic_action_view_as_list');
+    window.addHook('paneChanged', function (id) {
+      if (id == 'debug') {
```
Comment (johnd0e) [inline plugins/debug-console.js]: eqeqeq

**Review by johnd0e — CHANGES_REQUESTED** [review]

**Comment by johnd0e** [comment]
conflicts!

File: plugins/debug-console.js (hunk 10 above)
Comment (le-jeu) [inline plugins/debug-console.js]:
```suggestion
  if ($('#chatcontrols .active').text() !== 'debug')
```

File: plugins/debug-console.js (hunk 12 above)
Comment (le-jeu) [inline plugins/debug-console.js]:
```suggestion
      if (id === 'debug') {
```

File: plugins/debug-console.js (hunk 7 above)
Comment (le-jeu) [inline plugins/debug-console.js]:
```suggestion
      if (nativeConsole) {
        nativeConsole[which].apply(nativeConsole, arguments);
      }
```

File: plugins/debug-console.js (hunk 9 above)
Comment (le-jeu) [inline plugins/debug-console.js]:
```suggestion
        if ($('#chatcontrols .active').text() === 'debug') {
```

**Review by modos189 — APPROVED** [review]

**Pull request merged**